subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,742 @@
+#include <zephyr/net/coap_client.h>
+
+#define COAP_VERSION 1
+#define COAP_PATH_ELEM_DELIM '/'
+#define COAP_PATH_ELEM_QUERY '?'
+#define COAP_PATH_ELEM_AMP '&'
+#define COAP_SEPARATE_TIMEOUT 6000
+#define DEFAULT_RETRY_AMOUNT 5
+#define BLOCK1_OPTION_SIZE 4
+#define PAYLOAD_MARKER_SIZE 1
+
+static int coap_client_schedule_poll(struct coap_client *client, int sock,
+				     struct coap_client_request *req)
+{
+	client->fd = sock;
+	client->coap_request = req;
+
+	k_sem_give(&client->coap_client_recv_sem);
+	atomic_set(&client->coap_client_recv_active, 1);
+
+	return 0;
+}
+
+static int send_request(int sock, const void *buf, size_t len, int flags,
+			const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	if (addrlen == 0) {
+		return sendto(sock, buf, len, flags, NULL, 0);
+	} else {
+		return sendto(sock, buf, len, flags, dest_addr, addrlen);
+	}
+}
+
+static int receive(int sock, void *buf, size_t max_len, int flags,
+		   struct sockaddr *src_addr, socklen_t *addrlen)
+{
+	if (*addrlen == 0) {
+		return recvfrom(sock, buf, max_len, flags, NULL, NULL);
+	} else {
+		return recvfrom(sock, buf, max_len, flags, src_addr, addrlen);
+	}
+}
+
+static void reset_block_contexts(struct coap_client *client)
+{
+	client->recv_blk_ctx.block_size = 0;
+	client->recv_blk_ctx.total_size = 0;
+	client->recv_blk_ctx.current = 0;
+
+	client->send_blk_ctx.block_size = 0;
+	client->send_blk_ctx.total_size = 0;
+	client->send_blk_ctx.current = 0;
+}
+
+static int coap_client_init_path_options(struct coap_packet *pckt, const char *path)
+{
+	int ret = 0;
+	int path_start, path_end;
+	int path_length;
+	bool contains_query = false;
+	int i;
+
+	path_start = 0;
+	path_end = 0;
+	path_length = strlen(path);
+	for (i = 0; i < path_length; i++) {
+		path_end = i;
+		if (path[i] == COAP_PATH_ELEM_DELIM) {
+			/* Guard for preceding delimiters */
+			if (path_start < path_end) {
+				ret = coap_packet_append_option(pckt, COAP_OPTION_URI_PATH,
+								path + path_start,
+								path_end - path_start);
+				if (ret < 0) {
+					LOG_ERR("Failed to append path to CoAP message");
+					goto out;
+				}
+			}
+			/* Check if there is a new path after delimiter,
+			 * if not, point to the end of string to not add
+			 * new option after this
+			 */
+			if (path_length > i + 1) {
+				path_start = i + 1;
+			} else {
+				path_start = path_length;
+			}
+		} else if (path[i] == COAP_PATH_ELEM_QUERY) {
+			/* Guard for preceding delimiters */
+			if (path_start < path_end) {
+				ret = coap_packet_append_option(pckt, COAP_OPTION_URI_PATH,
+								path + path_start,
+								path_end - path_start);
+				if (ret < 0) {
+					LOG_ERR("Failed to append path to CoAP message");
+					goto out;
+				}
+			}
+			/* Rest of the path is query */
+			contains_query = true;
+			if (path_length > i + 1) {
+				path_start = i + 1;
+			} else {
+				path_start = path_length;
+			}
+			break;
+		}
+	}
+
+	if (contains_query) {
+		for (i = path_start; i < path_length; i++) {
```

Comment:
`i` is undefined

Reply:
it is defined in the beginning of the function.

Reply:
sorry, I placed this at the wrong point in the code